File: 35. 翻转链表.cpp
/**
 * Definition of singly-linked-list:
 *
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *        this->val = val;
 *        this->next = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param head: n
     * @return: The new head of reversed linked list.
     */
    ListNode * reverse(ListNode * head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;
        while (curr != nullptr) {
            ListNode* temp_next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp_next;
        }
        return prev;
    }
};

// 总耗时 82 ms
// 您的提交打败了 86.80% 的提交!

File: Python3/35. 翻转链表.py
"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""

class Solution:
    """
    @param head: n
    @return: The new head of reversed linked list.
    """
    def reverse(self, head):
        prev = None
        while head :
            temp, head.next = head.next, prev
            prev, head = head, temp
        return prev

# 总耗时 1007 ms
# 您的提交打败了 63.75% 的提交!